universal_transfer_operator/src/universal_transfer_operator/data_providers/filesystem/aws/s3.py

```diff
@@ -0,0 +1,151 @@
+)
+from universal_transfer_operator.datasets.base import UniversalDataset as Dataset
+from universal_transfer_operator.universal_transfer_operator import TransferParameters
+
+
+class S3DataProvider(BaseFilesystemProviders):
+    """
+    DataProviders interactions with S3 Dataset.
+    """
+
+    def __init__(
+        self,
+        dataset: Dataset,
+        transfer_params: TransferParameters = attr.field(
+            factory=TransferParameters,
+            converter=lambda val: TransferParameters(**val) if isinstance(val, dict) else val,
+        ),
+        transfer_mode: TransferMode = TransferMode.NONNATIVE,
+    ):
+        super().__init__(
+            dataset=dataset,
+            transfer_params=transfer_params,
+            transfer_mode=transfer_mode,
+        )
+        self.transfer_mapping: set = {
+            Location.S3,
+            Location.GS,
+        }
+
+    @cached_property
+    def hook(self) -> S3Hook:
+        """Return an instance of the database-specific Airflow hook."""
+        return S3Hook(
+            aws_conn_id=self.dataset.conn_id,
+            verify=self.verify,
+            transfer_config_args=self.transfer_config_args,
+            extra_args=self.s3_extra_args,
+        )
+
+    def check_if_exists(self) -> bool:
+        """Return true if the dataset exists"""
+        return self.hook.check_for_key(key=self.dataset.path)
+
+    @contextmanager
+    def read(self) -> list[TempFile]:
+        """Read the file from dataset and write to local file location"""
+        if not self.check_if_exists():
+            raise ValueError(f"{self.dataset.path} doesn't exits")
+        files = self.hook.list_keys(
+            bucket_name=self.bucket_name,
+            prefix=self.s3_key,
+            delimiter=self.delimiter,
+        )
+        local_file_paths = []
+        try:
+            for file in files:
+                local_path = self.hook.download_file(key=file)
+                local_file_paths.append(local_path)
+            yield local_file_paths
+        finally:
+            # Clean up the local files
+            self.cleanup(local_file_paths)
+
+    def write(self, source_ref: list[TempFile]):
+        """Write the file from local file location to the dataset"""
+
+        dest_s3_key = self.dataset.path
+
+        if not self.keep_directory_structure and self.prefix:
+            dest_s3_key = os.path.join(dest_s3_key, self.prefix)
+
+        destination_keys = []
+        if source_ref:
+            for file in source_ref:
+                if os.path.exists(file.tmp_file.name):
+                    dest_key = os.path.join(dest_s3_key, os.path.basename(file.actual_filename))
+                    self.hook.load_file(
+                        filename=file.tmp_file.name,
+                        key=dest_key,
+                        replace="replace",
+                        acl_policy=self.s3_acl_policy,
+                    )
+                    destination_keys.append(dest_key)
```

Comment:
This level of nesting is hard to read. Can you separate this into a helper function?

Reply:
@sunank200 it would be nice if we somehow could use _pathlib_ here. This would simplify the logic a lot imo. 

For example, something like (pseudo-code):
```python
            for file in source_ref:
                if file.tmp_file.exists():  # checks if the pathlib.Path obj exists
                    self.hook.load_file(
                        filename=file.tmp_file.as_posix(),  # returns pathlib.Path str representation
                        key=dest_s3_key / file.actual_filename.name,  # join path using pathlib "/" operator; `.name` on Path obj returns the last element of the path
                        replace="replace",
                        acl_policy=self.s3_acl_policy,
                    )
                    destination_keys.append(dest_key)
```

Reply:
https://github.com/astronomer/astro-sdk/pull/1492/commits/29e5ffaf685e27c9b5da3db1659b3442f78f68c9 and https://github.com/astronomer/astro-sdk/pull/1492/commits/535465e155915816db932fb678a30687a400e304